yaze.1
.TH YAZE-AG 1 2022-02-16 2.51.3 "Frank, John Saxton and AGL Hacks"
.SH NAME
yaze\-ag \- yet another z80 emulator by ag
.SH SYNOPSIS
.PU
.ll +8
.B yaze
.RB [ -1 ]
.RB [ -v ]
.RB [ -b\fIbootfile ]
.RB [ -l\fIloadadr ]
.RB [ -p\fIbasepage ]
.RB [ -s\fIstartup ]
.RB [ -z\fIZ3ENV ]
.I command...
.ll -8
.br
.B cpm
.I ...
.br
.B z80
.I ...
.br
.SH DESCRIPTION
.B yaze-ag
has been designed to provide an exact simulation of the Z80 microprocessor
on a Unix system. In addition to the instruction set emulator, a
CP/M basic i/o system is provided that can reference Unix directories
and Unix files containing images of CP/M disks.  When a genuine CP/M
operating system (ccp + bdos) or a clone system is loaded into the
emulator, a complete CP/M system is available, running on a
.br
Unix-/Linux-/RaspberryPi-/Mac-/Windows-Host.

.B yaze-ag-2.51.3
is a further development of yaze-1.10. New in yaze-ag is
the emulated Memory Management Unit (MMU) and the BIOS that supports
CP/M 3.1 or ZPM3 (boot file yaze-cpm3.boot) which is used in yaze-ag.
You can still run the CP/M 2.2 replacement "SuperDos" (boot file
yaze.boot) in yaze-ag if you want.
.br

New in version
.B 2.40.5
was the key translation module. So you can use e.g. the cursor
keys and function keys in the CP/M command line or when Wordstar
is running. Look at the files $HOME/cpm/*.ktt on the host.
.br

New in version
.B 2.51.1
was the possibility to define a clock rate. This allows
you to run the emulator e.g. with 4.0 MHz.
.br
Further there is a "window size handler" in the binary of yaze-ag. This
handler determines the coordinates of the used Terminal emulator (e.g. xterm,
gnome-terminal, mintty (Cygwin/Windows), ...). So the lines and columns
variables in the System Control Block (SCB) of CP/M-3 are set by this handler.
The patched Wordstar 4.00 (see drive A:) and the CP/M-3-Utilities
(e.g. TYPE.COM, 3DIR.COM, DIR.COM, ...) uses this two variables
in the SCB via the standardized BDOS3-Interface (call to 0005H) to
setup the screen.
.br
Further you can test the speed of the Z80 core with the submit file
"A:TESTZEX.SUB". It runs ZEXALL.COM and measure up the time how it's need and
prints the speed at the end of the submit file.
.br

New in version
.B 2.51.2
is the Macro Assembler "ZSM4". It's a Z80/Z180/Z280 Assembler. The licence is
the GPLv2 (GNU GENERAL PUBLIC LICENSE Version 2, June 1991).
The project page is "https://github.com/hperaza/ZSM4". The source files
of this Assembler can be mounted on drive O:. This Assembler is now the
standard for assemble BOOTSYS and the BIOS3-Z80-files for YAZE-AG
(see drive P:).
.br
Further with the flag "-?" you can print the usage, the features of yaze-ag
and at last it prints the copyright report.

New in version
.B 2.51.3
is "BOOTSYS version 2.00". When you call BOOTSYS.COM it tries to open
A:CPM3.SYS.
Where A: is the drive which is defined as the default drive in BOOTSYS.MAC
(see drive P:).
.br
You can define a drive as a parameter (e.g. "bootsys p:" (the colon ist
necessary)). BOOTSYS tries to open CPM3.SYS on that drive.
After load of CPM3.SYS
it relocates CPM3.SYS and starts CP/M 3.1 (cold boot).
.br
An another way to load CPM 3.1 is to generate a CPM3.COM with the following
command:

             A>pip cpm3.com=bootsys.com,cpm3.sys

You simple append CPM3.SYS after BOOTSYS.COM and save both in CPM3.COM.
When you call CPM3.COM BOOTSYS tests if a CPM3.SYS is appended and relocates
at once CPM3.SYS and starts CP/M 3.1 (cold boot).
.br
The boot file "yaze-cpm3.boot" which is load when you start YAZE-AG is
identical with A:CPM3.COM.

.B yaze
is a shell script. It is used to start yaze-ag.
It first tests if the startup file
.B .yazerc
is in the
current directory. If there is one, it immediately starts
.B yaze_bin
the binary of the Z80 emulator.
.br
If .yazerc does not exist, it tests if the directory $HOME/cpm exists.
If not, it creates $HOME/cpm. Then it copies some yaze-disks-files (.ydsk)
to $HOME/cpm
which are necessary for running yaze-ag with CP/M 3.1 or ZPM3.
Then it creates $HOME/cpm/doc_html and sets links to the
complete documentation in this directory.
This only happens if the directory $HOME/cpm does not exists.
.br
Finally, it changes to $HOME/cpm and starts
.B yaze_bin.
This is done every time you start the shell script yaze and if .yazerc
is not present in the current directory.

yaze\-ag can also be started with
.B z80
or
.B cpm.
They are symbolic links to the shell script yaze.

The older documentation is contained in the file yaze.doc (version yaze-1.10)
and the new features until yaze-ag-2.51.3 describes the file yaze-ag.doc
.br
(see $HOME/cpm/doc_html/index.html).
.SH OPTIONS
.TP
.B -?
prints the Usage, the Features and the Copyrights.
.TP
.B -1
set sector size to 128 bytes for all disks (only CP/M 3.1). 

If you create a disk file under the yaze-ag monitor or with the cdm(1)
utility and you use the default sectors per track (see create) the
sector size is also set to 2048 bytes (only CP/M 3.1).

If you use software like a disk edit utility under CP/M 3.1 it can be
necessary to set the sector size to 128 bytes.

See also the monitor command
.B 128
above in the section MONITOR COMMANDS.
.TP
.B -v
causes a summary of the system configuration to be displayed after
loading.
.TP
.B -b
gives the name of a file to be loaded into the emulated processor's
ram before the emulation is started.  The file is loaded at the
address given by the \fB-l\fR option, if one is present, or else at
the basepage (see \fB-p\fR).  If a boot file is not specified with the
\fB-b\fR option, yaze-ag looks firstly for the file yaze-cpm3.boot in
the current directory, then in /usr/local/lib/yaze to boot CP/M 3.1.
If yaze-ag is compiled without the flag -DBOOTSYS yaze-ag looks for the
file yaze.boot (instead of yaze-cpm3.boot) to boot the CP/M 2.2 replacement
Superdos.
.TP
.B -l
the hexadecimal address at which to load and start the bootstrap file,
if a separate bootstrap is necessary.

If you want to run CP/M 2.2 while yaze-ag is compiled to load yaze-cpm3.boot
(CP/M 3.1) you can start yaze-ag with the following options:

.B yaze -l -1 -b yaze.boot
.TP
.B -p
the top 2 hex digits of the location of the CP/M console command
processor (CP/M 2.2).  Also the location to which boot file is loaded if a
\fB-l\fR option is not present.
.TP
.B -s
a file containing monitor commands that are to be executed before
starting the emulator.  If no \fB-s\fR option is present, .yazerc from
the current directory or from the user's home directory ($HOME/cpm) is taken.
.TP
.B -z
the hexadecimal address in the emulated processor's ram of a 1KByte
area that should be reserved for use by an extended CP/M clone such as
ZCPR3.
.SH MONITOR COMMANDS
When the emulator is started it executes commands first from the
startup file \fB".yazerc"\fR, then from the tail of the command line
(any strings
which follow the options) and then from the tty if neither the startup
file nor the command line included a "go" command.  Control can be
returned to the monitor by executing the CP/M sys.com command or, if
yaze was compiled with -DDEBUG, by sending it a SIGINT signal.

Command names can be shortened to unique abbreviations.  If
\fByaze-ag\fR was compiled with GNU Readline, command-line editing,
filename completion and history recall are available - see the
readline documentation.

The monitor commands are described with more details in the file
yaze.doc. The new commands '128' and 'clock' are described in yaze\-ag.doc.
.br
.TP
.B help
Display a command list
.TP
.B help \fIcmd\fR
Give details about \fIcmd\fR
.TP
.B attach
Attach CP/M device to a Unix file/device
.TP
.B clock
Display and set the clock rate

Without a parameter it displays the current clock rate. With a parameter (for
example 4000 for 4.0 MHz) the clock rate will be set. The parameter "0" returns
to "max speed".
.br
If you want to set a permanent clock rate you can put a clock command
into the yaze-ag startup file $HOME/cpm/.yazerc.
.TP
.B detach
Detach CP/M device from file
.TP
.B setaccess
Turns on/off access time stamps for mounted directories
.TP
.B mount
Mount a Unix file or directory as a CP/M disk
.TP
.B remount
Remount a CP/M disk
.TP
.B umount
Unmount a CP/M disk
.TP
.B create
Create a new disk
.TP
.B 128
Set sector size to 128 for all disks (only CP/M 3.1).
To reverse this option you have to restart yaze-ag.
.TP
.B interrupt
Set user interrupt key
.TP
.B go
Start/Continue CP/M execution
.TP
.B !
Execute a Unix command
.TP
.B quit
Terminate yaze
.TP
.B time
Display elapsed time since last `time' command
.SH SUPPORTED OPERATING SYSTEMS
Yaze-ag-2.51.3 can be compiled under all unix or unix-like operating
systems.

yaze-ag-2.51.3 are tested with:
.TP
.B Solaris
cc and gcc (SparcV9 and X86_64)
.TP
.B Linux
gcc (x86_64, x86_32, ARM6, ARM7 and ARM8)
.TP
.B Raspberry Pi OS (Linux, Debian)
Raspberry Pi 1          (ARM1176JZF-S, single core, 32 bit)
.br
Raspberry Pi 2 model B  (Cortex-A7, quad core, 32 bit)
.br
Raspberry Pi 3 model B  (Cortex-A53, quad core, 32 & 64 bit)
.br
Raspberry Pi 3 model B+ (Cortex-A53, quad core, 32 & 64 bit)
.br
Raspberry Pi 4 model B  (Cortex-A72, quad core, 32 & 64 bit)
.TP
.B FreeBSD
clang (OPNsense)
.TP
.B MacOS-X
gcc ?
.TP
.B Cygwin
gcc (have a look to www.cygwin.com)
.br
Cygwin is a unix-like environment for windows.
.br
The "Standalone Binaries for Windows" are produced with Cygwin.
.SH SOURCE AND BINARIES
You find the source (for all systems) and the binaries for windows (if
you don't want to install the cygwin environment) with following URLs:

http://yaze-ag.de

or

http://www.mathematik.uni-ulm.de/users/ag/yaze-ag/
.SH SEE ALSO
.B cdm(1)
.SH AUTHORS
.TP
.B Frank D. Cringle
(E-Mail: fdc (at) cliwe (dot) ping (dot) de)
.br
(Copyright (c) 1995,1988 until yaze-1.10, CP/M 2.2)
.TP
.B Jon Saxton
He passed at June 15, 2015 after a tough battle with cancer.
.br
(Keytranslation Copyright (c) 2000,2015 by Jon Saxton)
.TP
.B Andreas Gerlich
(E-Mail: development (at) yaze-ag (dot) de)
.br
(MMU, CP/M 3.1 and other  extensions Copyright (c) 2000,2022 by
Andreas Gerlich, since yaze-1.10 until version yaze-ag-2.51.3)
.SH LICENSE
The License of this Open Source Project is the GPLv2 (GNU GENERAL
PUBLIC LICENSE Version 2, June 1991). For details see the file COPYING.


ChangeLog
Sorry, my english language is not good, I hope you understand what I write here.


2008-06-06  Andreas Gerlich

	pre release 2.20-RC1 (ready for Unix/Linux/FreeBSD)

	* I complete the documentation
	  (the use of the serial lines is now documented)
	  (I add in the HTML-Documentation the contens of
	   the CP/M 3 helpfile "HELP.HLP")

	* I change/complete the install process (make install)
	  and the startup shell script "yaze.sh"


2008-06-02  Andreas Gerlich

	development version 2.13-12

	* fix a severe bug when connecting (Unix-/Windows-)directories
	  to a CP/M drive.


2008-03-29  Andreas Gerlich

	development version 2.13-11

	* you can use COM1 (/dev/ttyS0) and COM2 (/dev/ttyS1)
	  the seriell lines to an another computer.
	  (not yet documented)

-------------------------------------------------------------------------------

2008-03-29  Andreas Gerlich

	development version 2.13-11

	* you can use COM1 (/dev/ttyS0) and COM2 (/dev/ttyS1)
	  the seriell lines to an another computer.
	  (not yet documented)

2005-10-18  Andreas Gerlich

	final release 2.12.1

	* Bugfix: Jim Donoghue (USA, Houston, Texas) reports a bug in the
	  Z80 assembler instructions OUTI and OUTD.

2005-08-21  Andreas Gerlich

	final release 2.12

	* the "Umlaute" are handled correct now
	  (this is only important for the Windows version of yaze-ag)


2005-07-16  Andreas Gerlich

	final release 2.11

	* some changes in the Makefiles

	I produce also the Windows binaries (build 01)


2005-07-05  Andreas Gerlich

	pre release 2.10-pre

	Windows Versions have now the "build"-counter

	* UCSD-Mode to run the UCSD P-System

	* bios routines conin and conout do not mask with 0x7F

	* I add the "DISKSORT" project

	This version is for testing !!!

	(read also all infomations until 2003-12-20 in this document.
	 At 2003-12-20 I begin with this development circle)


2005-03-05  Andreas Gerlich

	development Version 2.02-11

	* I adapt the emulator for the CYGWIN-Environment
	  (remember CYGWIN is an unix environment for windows)

	This version is for testing !!!

	(read also all infomations until 2003-12-20 in this document.
	 At 2003-12-20 I begin with this development circle)


2004-08-16  Andreas Gerlich

	development Version 2.02-10

	* NEW: P:BIOSKRNL.Z80 changed. Now you can run Z3PLUS !!!  :-)

	* NEW: I have to change the mount.com utility.
	       Please don't use the old one (yaze crashes) !!!

	* NEW: I put the changed documentation to the archive.
	       (It's the documentation of the final release 2.10)

	This version is for testing !!!

	(read also all infomations until 2003-12-20 in this document.
	 At 2003-12-20 I begin with this development circle)


2004-04-14  Andreas Gerlich

	development Version 2.02-08

	* NEW: Now the time stamps informations are now correct on big endian
	  (sparcv9) and on little endian (intel) machines.

	* NEW: I integrate the new monitor command "setaccess on/off".
	  (see also "help setaccess" in the monitor)
	  When a Unix directory is connected to a CP/M drive normaly
	  only update time stamps are present.
	  If you will see also the access time stamps you switch
	  it on with "setaccess on".
	  (see "SET [ACCESS=ON]" in the help with HELP.COM)

	This version is for testing !!!

	(read also all infomations until 2003-12-20 in this document.
	 At 2003-12-20 I begin with this development circle)


2004-01-27  Andreas Gerlich

	development Version 2.02-06

	* NEW: When you connect a directory to a CP/M drive the date stamps
	  of the unix files will now be present also under CP/M 3.1.
	  The modification time of the unix files is the update time
	  under CP/M 3.1 .

	  Now it's posible to edit source files under unix/linux and
	  to compile it under the yaze-ag emulator with the MAKE.COM
	  (and resetdsk.com, look version 2.02-02 ) utility.

	>>>> The time stamps are only correct on little endian machine
		(for example Intel) !!!

	This version is for testing.


2004-01-25  Andreas Gerlich

	development Version 2.02-04 (internal version not uploaded)

	* integrate the date calculation function "dayfaktor" in bios.c
	  (It calculates the theoretical days since 1.1.0000)
	  (Gregorian calendar)
	  (look to the Modula-2 sources DAYS.MOD/DEF and BIRTHDAY.MOD
	   on drive M: (disk file "Turbo-Modula-2") when you have started
	   the emulator.)

	* Update the CP/M 3.1 function "Get and Set Time" in bios.c to use
	  "dayfaktor".

	* patch cdm.c with the patch from Carl Mascott.

	* Carl Mascott reports an error with unmounting drives.
	  The sys.com utility makes now a full reset of all drives (BDOS 37).

	* cleanup all sources (*.c) to use the flag -D_BSD_SOURCE for Linux
	  (I took the knowledge about this flag from yaze-1.12)


2004-01-20  Andreas Gerlich

	development version 2.02-02

	* Fix reread bug of a unix directory

	  Describtion:
	  When you give a ^C (CTRL-C) at the prompt of the CCP
	  the disk system in CP/M will be reseted completely.
	  A drive which is connected to a directory will be
	  now fully rereaded (remount) at the next access (SELDSK).

	  You can use the feature "Mounting Directories" if you create
	  and edit files with vi or an another editor under unix and
	  you want to compile the files over a compiler/assembler in yaze-ag.
	  (This new function works also under superdos (CP/M 2.2) which is
	  in the boot file "yaze.boot".)

	  I add the little program "RESETDSK.Z80/COM" which resets all
	  disks over the BDOS function 37. This is identical like the
	  CTRL-C. You can use this utility in a submit file:

		compile.sub:

		;
		resetdsk
		;
		; Resets all disks. Every drive will be rereaded by
		; CP/M 3.1. Direcrories will be fully rereaded.
		;
		m80 d:test=c:test
		;
		; d: must be a disk file.
		; c: is connected to a directory and will be fully rereaded

	  I add to the monitor the command "remount <drive>" which you can use
	  also in a submit file with the command "sys remount <drive>"

yaze.c
/* yaze-ag - yet another Z80 emulator by ag.
   Copyright (C) 1995,1998  Frank D. Cringle.
   Modifications for CP/M 3.1 Copyright (C) 2000/2004 by Andreas Gerlich (agl)

Yaze-ag is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
/* #include <stdbool.h> */

#include "ytypes.h"

#include "mem_mmu.h"
#include "simz80.h"
#include "yaze.h"

/* Z80 registers */
WORD af[2];			/* accumulator and flags (2 banks) */
int af_sel;			/* bank select for af */

struct ddregs regs[2];		/* bc,de,hl */
int regs_sel;			/* bank select for ddregs */

WORD ir;			/* other Z80 registers */
WORD ix;
WORD iy;
WORD sp;
WORD pc;
WORD IFF;

/* definitiions for the brake */
FASTREG tStates_save;
FASTREG clockFrequency_save = 0;	/* default no brake */
FASTREG startTime_save;
FASTREG tStatesInSlice_save;
bool brakeini = true;		/* must be true! */

/* See definitions for memory in mem_mmu.[c,h] (agl) */

#ifndef LIBDIR
#define LIBDIR "/usr/local/lib/yaze"
#endif

#ifdef BOOTSYS
static char *bootfile = "yaze-cpm3.boot";
#else
static char *bootfile = "yaze.boot";
#endif
static char *startup = ".yazerc";
static char *progname;
#ifdef BOOTSYS
static long loadadr = 0x100;
#else
static long loadadr = -1;
#endif
#ifdef BIOS
static long basepage = 0xe4;
static int vflag;
int always128 = 0;	/* is set if all drives uses 128 byte sectors	*/
			/* only used under CP/M 3.1			*/
long z3env;

WORD	ccp_base;		/* base address of ccp */
WORD	bdos_base;		/* base address of bdos */
WORD	bios_base;		/* base address of bios */
WORD	bios_top;		/* end of bios, start of global work area */

WORD	dirbuff;		/* common directory buffer for all disks */
WORD	dptable;		/* base of disk parameter table */
BYTE	*global_alv;		/* global allocation vector */
#endif

extern bool rtc_avail; /* nur zum Testen */

static void
usage(void)
{
    char c;

    fprintf(stderr,
            "Usage: %s {flags} {commands}\n"
            "   -b <file>       boot from this file\n"
            "   -l <xxxx>       load bootfile at given (hex) address\n"
#ifdef BIOS
            "   -p <xx>         base page (top 2 hex digits of ccp base)\n"
#endif
            "   -s <file>       startup file\n" ,progname);

#ifdef BIOS
    fprintf(stderr,
	    "   -h <n>          screen height (default 25)\n"
	    "   -w <n>          screen width (default 80)\n"
            "   -v              verbose\n"
            "   -z <xxxx>       (hex) address of z3env (if desired)\n"
            "   -1              all drives uses 128 byte sectors (under CP/M 3.1)\n"
            "                   (normal 2048 byte sectors are used, 128 byte\n"
            "                   sectors are necessary to use a disk editor)\n\n"
	    "Press RETURN to continue ...");
    read(fileno(stdin), &c, 1);
    fprintf(stderr,
	    "\n"
	    "Features: MMU:     Memory Management Unit with 16 tables and 16 page pointers\n"
	    "   of   :          per table. This is the basis for running CP/M 3.1. A bank\n"
	    "YAZE-AG :          switch is simple done with a pointer to one of the 16 tables\n"
	    "                   and is very fast.\n\n"
	    "          KeyTr:   Key translation: Allows to use the cursor keys of the PC in\n"
	    "                   Wordstar and on the command line of CP/M 3.1. Commands can\n"
	    "                   be put on function keys and many more. This software module\n"
	    "                   is Copyright (c) by John Saxton (see yaze-keys.pdf).\n\n"
	    "          WinSz:   Window size handler/routine: This handler initializes the\n"
	    "                   lines and columns in the SCB of CP/M 3.1 like the lines and\n"
	    "                   columns of the used terminal emulator. The CP/M 3.1 system\n"
	    "                   utilities (like TYPE, DIR, ...) and the patched Wordstar\n"
	    "                   4.00 (A:WS.COM) uses this two variables to setup the screen.\n\n"
	    "          MHz:     You can define a clock rate in the monitor of yaze-ag.\n"
	    "                   This allows to run the emulator at the defined clockrate\n"
	    "                   like e.g. \"4000\" for 4000 kHz (4.0 MHz) or any other clock\n"
	    "                   rate. Clock rate \"0\" returns to max speed.\n\n"
	    "Press RETURN to continue ...");
    read(fileno(stdin), &c, 1);
    fprintf(stderr,
	    "\n"
	    "          ZSM4:    ZSM4 is a free Macro Assembler for the Z80/Z180/Z280. The\n"
	    "                   Assembler is created by Hector Peraza. The project page is\n"
	    "                   on \"https://github.com/hperaza/ZSM4\". The licence is the\n"
	    "                   GPLv2 (General Public Licence Version 2). The source files\n"
	    "                   of this Assembler can be mounted on drive O:.\n"
	    "                   This Assembler is now the standard for assemble BOOTSYS and\n"
	    "                   the BIOS3-Z80-files for YAZE-AG (see drive P:).\n\n"
	    "Copyright (c) 1995,1998 by Frank D.Cringle (until yaze-1.10, CP/M 2.2).\n\n"
	    "Keytranslation is Copyright (c) 2000,2015 by Jon Saxton (Australia).\n\n"
	    "The basis for the MMU was the idea of the Pagetables by Michael Haardt.\n\n"
	    "MMU, CP/M 3.1 and other extensions Copyright (c) 2000,2022 by Andreas Gerlich.\n"
	    "(See the file yaze-ag_doc.html in the documentation about the news features of\n"
	    " yaze-ag-2.xx.x since yaze-1.10.)\n\n"
	    "The License of this Open Source Project is the GNU GENERAL PUBLIC LICENSE\n"
	    "Version 2, June 1991 (GPLv2).\n\n"
	
	    "Press RETURN to continue ...");
    read(fileno(stdin), &c, 1);
#endif

    exit (1);
}

/* This routine loads the bootstrap file into the simulated processor's ram.

   If a load address is given with the -l flag, the boot file is
   loaded to that address, the z80 pc is set to it and hl is set to
   the ccp base address.  The bootfile might be a standalone program
   which ignores the bios vectors or it might contain a
   self-relocating ccp/bdos (clone), in which case it should relocate
   itself and jump to the bios cold boot vector.

   Otherwise, the boot file can contain 1 or 2 copies of cp/m.  If it
   contains 1 copy the specified or default basepage must correspond
   to the run-time location of the cp/m image.  If it contains 2
   copies of the same code assembled to run at different locations,
   the code can and will be relocated to fit the basepage in use. */

static void
load_cpm(const char *bfname)
{
    FILE *cpm;
#ifdef BIOS
    int i, skip = 0;
    WORD offs;
    struct stat	sb;
#endif

    if ((cpm = fopen(bfname, "r")) == NULL)
    {
	char *p = xmalloc(strlen(bfname) + strlen(LIBDIR) + 1);
	strcpy(p, LIBDIR);
	strcat(p, bfname);
	if ((cpm = fopen(p, "r")) == NULL)
	    free(p);
	else
	    bfname = p;
    }
    if (cpm == NULL)
    {
	perror(bfname);
	exit(1);
    }
    if (loadadr >= 0)
    {
	fread(ram+loadadr, 1, 64*1024-loadadr, cpm);
	fclose(cpm);
	pc = loadadr;
#ifdef BIOS
	regs[regs_sel].hl = ccp_base;
#endif
	printf("Running '%s'\n", bfname);
	return;
    }
#ifdef BIOS
    if (fstat(fileno(cpm), &sb) != 0)
    {
	perror(bfname);
	exit(1);
    }
    if ((sb.st_size >= CPM_LENGTH+0x880 && sb.st_size < 2*CPM_LENGTH) ||
	(sb.st_size >= 2*CPM_LENGTH+0x880))
    {
	skip = 0x880;
	fseek(cpm, skip, SEEK_SET);
    }
    fread(ram + ccp_base, CPM_LENGTH, 1, cpm);
    if (sb.st_size < 2*CPM_LENGTH) {
	/* single, non-relocatable copy */
	if (ram[ccp_base] == 0xc3 && (ram[ccp_base+2]&0xfc) != basepage)
	    printf("Warning: %s appears to need -p %2x\n", bfname,
		   ram[ccp_base+2]&0xfc);
    }
    else
    {
	if ((sb.st_size & 1) || ram[ccp_base] != 0xc3)
	{
	    fprintf(stderr, "%s does not appear to contain 2 cp/m images\n", bfname);
	    exit(1);
	}
	fseek(cpm, skip + sb.st_size/2, SEEK_SET);
	/* this loses if the initial jump points into the 2nd half of ccp */
	offs = (ccp_base >> 8) - (ram[ccp_base + 2] & 0xfc);
	for (i = 0; i < CPM_LENGTH; i++)
	{
	    if ((BYTE) getc(cpm) != ram[ccp_base + i])
		ram[ccp_base + i] += offs;
	}
    }
    fclose(cpm);
    printf("Running '%s'\n", bfname);
    PutBYTE(3, 0x95);				/* iobyte */
    PutBYTE(4, 0);				/* disk A: */
    pc = bios_base;
#endif
}

#ifdef BIOS
#define N_VEC	17			/* number of bios entry points */

static void
setvectors(void)
{
    int i;
    long cpm_brk = 64*1024;

    if (z3env == cpm_brk - 1024)	/* z3env at top of memory? */
	cpm_brk = z3env;
    cpm_brk -= 128;			/* space for directory buffer */
    dirbuff = cpm_brk;
    cpm_brk -= 16*(16+15);		/* space for 16 disk parameter headers
					   and disk parameter blocks */
    dptable = cpm_brk;

    if (basepage*256 + CPM_LENGTH + 6*N_VEC >= dptable) {
	fprintf(stderr, "Not enough space in simulated ram to accommodate base at %02lX\n",
		basepage);
	exit(1);
    }

    for (i = 0; i < N_VEC; i++) { /* build bios jump table */
	BYTE *p = ram + bios_base + 3*i;
	p[0] = 0xc3;		/* jump */
	p[1] = 3*(i+N_VEC);
	p[2] = (bios_base >> 8);
	p[3*N_VEC] = 0x76;	/* halt */
	p[3*N_VEC+1] = i;	/* function code */
	p[3*N_VEC+2] = 0xc9;	/* return */
    }
    bios_top = bios_base + 6*N_VEC;
    i = (dptable - bios_top + 7) / 8;
    global_alv = xmalloc(i);
    memclr(global_alv, i);
}
#endif

/* getopt declarations */
extern int getopt(int argc, char * const argv[],
		  const char *optstring);
extern char *optarg;
extern int optind, opterr, optopt;

/* Core Z80Core = simz80;	\* funktion pointer default simz80() */

int
main(int argc, char **argv)
{
    int
	c,
	gostart,
	term_height = 25,
	term_width  = 80;

#ifdef BIOS
    int goopt;
#endif

    progname = argv[0];

/*  printf("\033]0;YAZE-AG-%s-%s%s\007",VERSION,DEVVER,BUILD); */
    printf("\033]0;YAZE-AG-%s%s\007",VERSION,BUILD);

    puts("\nYet Another Z80 Emulator by AG, "
	"\033[32m"

/*	"final release " VERSION  */
	"development version " VERSION "-" DEVVER 
/*	"release candidate " VERSION "-" DEVVER */

	"\033[0m"

#ifdef UCSD
	"/UCSD"
#endif
/*
#ifdef CYGWIN
	"\033[33m"
#endif
	BUILD 
#ifdef CYGWIN
	"\033[0m"
#endif
*/
	/* " (devel version)" */
#ifdef MMU
	"\r\nFeatures: MMU,KeyTr,WinSz,MHz,ZSM4, Copyright (use the flag \"-?\" for details)."
#endif
	"\r\nYAZE-AG comes with ABSOLUTELY NO WARRANTY; for details "
	"see the file \"COPYING\".\r\n");

    while ((c = getopt(argc, argv, "b:l:s:h:w:"
#ifdef BIOS
                                           "p:vz:1"
#endif

                                           "?"       )) != EOF)
	switch (c)
	{
	case 'b':
	    bootfile = optarg;
	    break;
	case 'l':
	    loadadr = strtol(optarg, NULL, 16);
	    break;
	case 's':
	    startup = optarg;
	    break;
#ifdef BIOS
	case 'p':
	    basepage = strtol(optarg, NULL, 16);
	    break;
	case 'v':
	    vflag = 1;
	    break;
	case 'z':
	    z3env = strtol(optarg, NULL, 16);
	    break;
	case '1':
	    always128 = 1;
	    break;
#endif
	case 'h':
	    term_height = strtol(optarg, NULL, 10);
	    break;
	case 'w':
	    term_width = strtol(optarg, NULL, 10);
	    break;
	case '?':
	    usage();
	}

#ifdef MMU
    initMEM();	/* initialice memory */
    initMMU();  /* initialice MMU */
#endif

#ifdef BIOS
    ccp_base = basepage * 256;
    bdos_base = ccp_base + CCP_LENGTH;
    bios_base = bdos_base + BDOS_LENGTH;

    /* Plug screen height and width int SCB */
    ram[bios_base - 0x100 + 0xBA] = term_height - 1;
    ram[bios_base - 0x100 + 0xBC] = term_width - 1;

    if (z3env == bios_base)
	bios_base += 1024;
#endif

    load_cpm(bootfile);
#ifdef BIOS
    setvectors();

    if (vflag)
    {
	printf("bootfile:	%s\n", bootfile);
	if (loadadr >= 0)
	    printf("loadadr:	%4lx\n", loadadr);
	printf("startup file:	%s\n"
	       "basepage:	%2lx\n"
	       "ccp_base:	%4hx\n"
	       "bdos_base:	%4hx\n"
	       "bios_base:	%4hx\n"
	       "bios_top:	%4hx\n",
	       startup, basepage, ccp_base,
	       bdos_base, bios_base, bios_top);
	if (z3env)
	    printf("z3env:		%4lx\n", z3env);
	printf("dptable:	%4hx\n"
	       "dirbuf:		%4hx\n",
	       dptable, dirbuff);
    }
#endif

    gostart = bios_init(startup);

#ifdef BIOS
    for (goopt = 0; !goopt && optind < argc; optind++)
	goopt = docmd(argv[optind]);

    if (!gostart && !goopt)
	monitor(0);
#endif

#ifdef DEBUG
    do
    {
	FASTWORK adr = simz80(pc);
	if (adr & 0x10000)
	    monitor(adr);
	else
	    bios(ram[adr]);
    } while (1);
#elif !BIOS
    simz80(pc);
    fprintf(stderr,"HALT\n\r");
    fprintf(stderr,"PC   SP   IR   IX   IY   AF   BC   DE   HL   AF'  BC'  DE'  HL'\n\r");
    fprintf(stderr,"%04x %04x %04x %04x %04x %04x %04x %04x %04x %04x %04x %04x %04x\n\r",pc,sp,ir,ix,iy,af[af_sel],regs[regs_sel].bc,regs[regs_sel].de,regs[regs_sel].hl,af[1-af_sel],regs[1-regs_sel].bc,regs[1-regs_sel].de,regs[1-regs_sel].hl);
    exit(0);
#else
    do
    {
	pc = clockFrequency_save ? simz80_with_tStates(pc) : simz80(pc);
	bios(GetBYTE(pc));
    } while (1);
#endif
}

#ifndef USE_GNU_READLINE
void *
xmalloc(size_t size)
{
    void *p = malloc(size);

    if (p == NULL)
    {
	fputs("insufficient memory\n", stderr);
	exit(1);
    }
    return p;
}
#endif
